File: hbp100/core/pipeline.py
import logging
from typing import Dict, Any, Optional, List, Tuple
from dataclasses import dataclass, field

from hbp100.schemas.entity import Entity
from hbp100.schemas.decision import PrivacyDecision, DecisionType
from hbp100.extractors.manager import ExtractorManager
from hbp100.policy_engine.predictor import PrivacyPredictor
from hbp100.placeholders.generator import PlaceholderGenerator
from hbp100.placeholders.validator import PlaceholderValidator
from hbp100.placeholders.restore import PlaceholderRestorer
from hbp100.utils.logger import get_logger

logger = get_logger(__name__)


@dataclass
class PipelineResult:
    original_text: str
    masked_text: str
    metadata: Dict[str, str] = field(default_factory=dict)
    entities: List[Entity] = field(default_factory=list)
    decisions: List[PrivacyDecision] = field(default_factory=list)
    has_pii: bool = False

    def to_dict(self) -> Dict[str, Any]:
        return {
            "original_text": self.original_text,
            "masked_text": self.masked_text,
            "metadata": self.metadata,
            "entities": [e.to_dict() for e in self.entities],
            "decisions": [d.to_dict() for d in self.decisions],
            "has_pii": self.has_pii,
        }


class PrivacyPipeline:
    def __init__(
        self,
        model_path: Optional[str] = None,
        vectorizer_path: Optional[str] = None,
        lazy_load: bool = True,
    ):
        self.extractor_manager = ExtractorManager()
        self.predictor = PrivacyPredictor(
            model_path=model_path,
            vectorizer_path=vectorizer_path,
        )
        self.generator = PlaceholderGenerator()
        self.validator = PlaceholderValidator()
        self.restorer = PlaceholderRestorer()

        if not lazy_load:
            self.load_assets()

        self._initialized = True
        logger.info("Privacy pipeline initialized")

    def load_assets(self) -> bool:
        return self.predictor.load_assets()

    def process(
        self,
        text: str,
        intent: Optional[str] = None,
        return_decisions: bool = False,
    ) -> PipelineResult:
        if not text or not text.strip():
            raise ValueError("Input text cannot be empty")

        self.generator.reset_all()
        self.validator.reset()
        self.restorer.reset()

        logger.info(f"Processing text (length: {len(text)} chars)")

        entities = self._extract_entities(text)
        logger.info(f"Extracted {len(entities)} entities")

        if not entities:
            return PipelineResult(
                original_text=text,
                masked_text=text,
                metadata={},
                entities=[],
                decisions=[],
                has_pii=False,
            )

        decisions = self._predict_decisions(entities, text, intent)
        logger.info(f"Predicted {len(decisions)} decisions")

        masked_text, metadata = self._apply_masking(text, entities, decisions)
        logger.info(f"Masked {len(metadata)} entities")

        self.validator.update_allowed(metadata)
        self.restorer.update_metadata(metadata)

        has_pii = any(d.should_mask for d in decisions)
        decisions_to_return = decisions if return_decisions else []

        return PipelineResult(
            original_text=text,
            masked_text=masked_text,
            metadata=metadata,
            entities=entities,
            decisions=decisions_to_return,
            has_pii=has_pii,
        )

    def _extract_entities(self, text: str) -> List[Entity]:
        return self.extractor_manager.extract_all(text)

    def _predict_decisions(
        self,
        entities: List[Entity],
        original_text: str,
        intent: Optional[str] = None,
    ) -> List[PrivacyDecision]:
        if not self.predictor.is_loaded:
            logger.warning("Predictor not loaded, using fallback decisions")
            return self._fallback_decisions(entities)

        return self.predictor.predict_batch(entities, original_text, intent)

    def _fallback_decisions(self, entities: List[Entity]) -> List[PrivacyDecision]:
        return [
            PrivacyDecision(
                entity=entity,
                decision=DecisionType.MASK,
                confidence=0.50,
                context_string=f"fallback [SEP] {entity.type.value} [SEP] {entity.value}",
            )
            for entity in entities
        ]

    def _apply_masking(
        self,
        text: str,
        entities: List[Entity],
        decisions: List[PrivacyDecision],
    ) -> Tuple[str, Dict[str, str]]:
        decision_map = {d.entity: d for d in decisions}

        sorted_entities = sorted(entities, key=lambda e: e.start, reverse=True)

        masked = text

        for entity in sorted_entities:
            decision = decision_map.get(entity)
            if decision and decision.should_mask:
                placeholder = self.generator.generate(entity)
                masked = (
                    masked[: entity.start] + placeholder + masked[entity.end :]
                )

        metadata = self.generator.get_metadata()
        return masked, metadata

    def restore_placeholders(
        self,
        text: str,
        metadata: Optional[Dict[str, str]] = None,
    ) -> str:
        return self.restorer.restore(text, metadata)

    def validate_response(self, response: str, metadata: Dict[str, str]) -> Tuple[bool, Optional[str]]:
        return self.validator.validate(response, metadata)

    def reset(self):
        self.generator.reset_all()
        self.validator.reset()
        self.restorer.reset()
        logger.info("Pipeline reset")

    @property
    def is_loaded(self) -> bool:
        return self.predictor.is_loaded

File: hbp100/src/core/pipeline.rs
use crate::extractors::ExtractorManager;
use crate::placeholders::{PlaceholderGenerator, PlaceholderValidator, PlaceholderRestorer};
use crate::policy_engine::{ContextBuilder};
use crate::schemas::{Entity, PrivacyDecision, ProcessResult};
use std::collections::HashMap;
use log;

#[cfg(feature = "python-bridge")]
use crate::python_bridge::MLPredictor;

#[cfg(not(feature = "python-bridge"))]
pub struct MLPredictor {
    // Fallback when Python bridge is disabled
    threshold: f32,
}

#[cfg(not(feature = "python-bridge"))]
impl MLPredictor {
    pub fn new() -> Self {
        Self { threshold: 0.5 }
    }
    
    pub fn predict(&self, _context: &str) -> (bool, f32) {
        // Fallback: always mask with 0.5 confidence
        (true, 0.5)
    }
}

pub struct Pipeline {
    extractor_manager: ExtractorManager,
    generator: PlaceholderGenerator,
    validator: PlaceholderValidator,
    restorer: PlaceholderRestorer,
    context_builder: ContextBuilder,
    predictor: Option<MLPredictor>,
}

impl Pipeline {
    pub fn new() -> Self {
        Self {
            extractor_manager: ExtractorManager::new(),
            generator: PlaceholderGenerator::new(),
            validator: PlaceholderValidator::new(),
            restorer: PlaceholderRestorer::new(),
            context_builder: ContextBuilder::new(),
            predictor: None,
        }
    }
    
    #[cfg(feature = "python-bridge")]
    pub fn with_predictor(mut self, predictor: MLPredictor) -> Self {
        self.predictor = Some(predictor);
        self
    }
    
    #[cfg(not(feature = "python-bridge"))]
    pub fn with_predictor(mut self, _predictor: MLPredictor) -> Self {
        self.predictor = Some(MLPredictor::new());
        self
    }
    
    pub fn process(&mut self, text: &str, intent: Option<&str>) -> ProcessResult {
        if text.trim().is_empty() {
            return ProcessResult::new(text, text);
        }
        
        self.generator.reset();
        self.validator.reset();
        self.restorer.reset();
        
        log::info!("Processing text (length: {} chars)", text.len());
        
        // Extract entities
        let entities = self.extractor_manager.extract_all(text);
        log::info!("Extracted {} entities", entities.len());
        
        if entities.is_empty() {
            return ProcessResult::new(text, text);
        }
        
        // Predict decisions
        let decisions = self.predict_decisions(&entities, text, intent);
        log::info!("Predicted {} decisions", decisions.len());
        
        // Apply masking
        let (masked_text, metadata) = self.apply_masking(text, &entities, &decisions);
        log::info!("Masked {} entities", metadata.len());
        
        self.validator.update_allowed(metadata.keys().cloned().collect());
        self.restorer.update_metadata(metadata.clone());
        
        let has_pii = decisions.iter().any(|d| d.should_mask());
        
        ProcessResult {
            original_text: text.to_string(),
            masked_text,
            metadata,
            entities,
            decisions,
            has_pii,
        }
    }
    
    fn predict_decisions(
        &self,
        entities: &[Entity],
        original_text: &str,
        intent: Option<&str>,
    ) -> Vec<PrivacyDecision> {
        if let Some(predictor) = &self.predictor {
            predictor.predict_batch(entities, original_text, intent)
        } else {
            // Fallback: mask everything
            entities.iter()
                .map(|e| {
                    let context = self.context_builder.build(e, original_text, intent);
                    PrivacyDecision::new(
                        e.clone(),
                        crate::schemas::DecisionType::Mask,
                        0.5,
                        context,
                    )
                })
                .collect()
        }
    }
    
    fn apply_masking(
        &self,
        text: &str,
        entities: &[Entity],
        decisions: &[PrivacyDecision],
    ) -> (String, HashMap<String, String>) {
        use std::collections::HashMap;
        
        // Build decision map
        let decision_map: HashMap<&Entity, &PrivacyDecision> = decisions
            .iter()
            .map(|d| (&d.entity, d))
            .collect();
        
        // Sort entities by start position descending for in-place replacement
        let mut sorted_entities: Vec<&Entity> = entities.iter().collect();
        sorted_entities.sort_by(|a, b| b.start.cmp(&a.start));
        
        let mut masked = text.to_string();
        let mut metadata = HashMap::new();
        
        for entity in sorted_entities {
            if let Some(decision) = decision_map.get(&entity) {
                if decision.should_mask() {
                    let placeholder = self.generator.generate(entity);
                    let start = entity.start;
                    let end = entity.end;
                    masked.replace_range(start..end, &placeholder);
                    metadata.insert(placeholder, entity.value.clone());
                }
            }
        }
        
        (masked, metadata)
    }
    
    pub fn restore_placeholders(&self, text: &str) -> String {
        self.restorer.restore(text)
    }
    
    pub fn restore_with_metadata(&self, text: &str, metadata: HashMap<String, String>) -> String {
        self.restorer.restore_with_metadata(text, metadata)
    }
    
    pub fn validate_response(&self, response: &str) -> (bool, Option<String>) {
        self.validator.validate(response)
    }
    
    pub fn reset(&mut self) {
        self.generator.reset();
        self.validator.reset();
        self.restorer.reset();
        log::info!("Pipeline reset");
    }
}

impl Default for Pipeline {
    fn default() -> Self {
        Self::new()
    }
}
